### PR TITLE
[gpio] Refactor SAM signals API

### LINKED ISSUE #481: Discussion about Signals API
This is a break-out from a discussion in #159 

@henrikssn:
> A nit on the API here, why does the user needs to know about `Dm` and `Dp` when initializing a USB peripheral?
> 
> I previously made this `Usb::connect<GpioA24, GpioA25>()` which has the same static safety but does not require the reader to fully understand a 1000+ pages datasheet.
> 
> I have a similar comment about Uart, where the user currently is required to know which pins map to which sercom pads, which at least I have no chance doing without digging datasheet or looking at the generated code (I just wanted to print hello world!). We could similarly avoid that by supporting something like `Uart0::connect<GpioA24, GpioA25>()`.

@salkinium:

> Been there, done that. I have some very in-depth knowledge (*sigh*) about a number of issues with the signals API in general that needed to be solved:
> 
> 1. You need a clear mapping of _at least_ (Peripheral, Pin, Signal) to determine the correct bits to flip. In your case the Signal is given implicitly by argument position.
> 2. For the STM32F1 series, you can only remap groups of signals to overlapping groups of pins. Thus depending on the specific (Pin, Signal) group mapping at the callsite you must choose from multiple remap groups or static assert. Example from a generated file (`examples/blue_pill/blink/modm/src/modm/platform/gpio/connector_specialized.hpp:270`):
> 
> ```c++
> static_assert(/* check if group mapping exists */,
>     "This pin set contains conflicting remap groups!\nAvailable groups for Tim2 are:\n"
>     " Tim2 |    0    |    1    |    2    |    3    \n"
>     "------|---------|---------|---------|---------\n"
>     "  A0  | Ch1,Etr |         | Ch1,Etr |         \n"
>     "  A1  |   Ch2   |         |   Ch2   |         \n"
>     "  A2  |   Ch3   |   Ch3   |         |         \n"
>     "  A3  |   Ch4   |   Ch4   |         |         \n"
>     " A15  |         | Ch1,Etr |         | Ch1,Etr \n"
>     "  B3  |         |   Ch2   |         |   Ch2   \n"
>     " B10  |         |         |   Ch3   |   Ch3   \n"
>     " B11  |         |         |   Ch4   |   Ch4   \n"
> );
> ```
> 
> 1. You need a variable number of connect pins with some default behavior. For example, you can use SPI to bit-bang WS2812 LEDs, but you only need MOSI for that. Or perhaps you only need to read something with only MISO and SCK. This makes it difficult to use fixed Signal argument positions with defaults:
> 
> ```c++
> template<class Sck = GpioUnused, class Mosi = GpioUnused, class Miso = GpioUnused >
> Spi::connect() {}
> 
> Spi::connect<GpioUnused, GpioA1>(); // For Mosi only
> Spi::connect<GpioA0, GpioUnused, GpioA2>(); // for Sck, Miso only
> ```
> 
> 1. Due to lack of named arguments, you don't actually know what pin is connected to what signal, particularly with defaulted arguments. This is much clearer:
> 
> ```c++
> Spi::connect<GpioA1::Mosi>();
> Spi::connect<GpioA0::Sck, GpioA2::Miso>();
> ```
> 
> 1. It's unclear what and how many Signals exactly you need to connect for your use-case For example the FSMC can be configured at runtime in different connection modes, making it quite difficult to do this _without_ variadic templates. This is still type-safe, but not nothing was explicitly coded for this config:
> 
> ```c++
> GpioConnector<Peripheral::Fsmc,
>     Board::display::fsmc::D0::D0,
>     Board::display::fsmc::D1::D1,
>     Board::display::fsmc::D2::D2,
>     Board::display::fsmc::D3::D3,
>     Board::display::fsmc::D4::D4,
>     Board::display::fsmc::D5::D5,
>     Board::display::fsmc::D6::D6,
>     Board::display::fsmc::D7::D7,
>     Board::display::fsmc::D8::D8,
>     Board::display::fsmc::D9::D9,
>     Board::display::fsmc::D10::D10,
>     Board::display::fsmc::D11::D11,
>     Board::display::fsmc::D12::D12,
>     Board::display::fsmc::D13::D13,
>     Board::display::fsmc::D14::D14,
>     Board::display::fsmc::D15::D15,
>     Board::display::fsmc::Noe::Noe,
>     Board::display::fsmc::Nwe::Nwe,
>     Board::display::fsmc::A18::A18
> >::connect();
> ```
> 
> 1. Finally, you can define pin configuration defaults easily with our current API. For example, you can use UART with Tx or Rx or Tx+Rx, but this code works for all three cases, since GpioUnused is returned from `GetSignal<Tx>` if the signal does not exist:
> 
> ```c++
> template< template<Peripheral _> class... Signals >
> static void
> connect(Gpio::InputType InputTypeRx = Gpio::InputType::PullUp,
>         Gpio::OutputType OutputTypeTx = Gpio::OutputType::PushPull)
> {
>     using Connector = GpioConnector<Peripheral::{{ name }}, Signals...>;
>     using Tx = typename Connector::template GetSignal< Gpio::Signal::Tx >; // Find Tx signal in group
>     using Rx = typename Connector::template GetSignal< Gpio::Signal::Rx >; // Find Rx signal in group
>     static_assert(((Connector::template IsValid<Tx> and Connector::template IsValid<Rx>) and sizeof...(Signals) == 2) or
>                   ((Connector::template IsValid<Tx> or  Connector::template IsValid<Rx>) and sizeof...(Signals) == 1),
>                   "{{ name }}::connect() requires one Tx and/or one Rx signal!");
> 
>     Tx::setOutput(OutputTypeTx); // Does nothing if GpioUnused
>     Rx::setInput(InputTypeRx); // Does nothing if GpioUnused
>     Connector::connect();
> }
> ```
> 
> What would be useful is a slim HTML-only version of a pinout picker, where you use modm-devices data to configure the signal<->pin map visually and then generate the appropriate modm code. That would be much easier to prototype complex dependencies like SystemClock configurations.
> 
> There's already a curses-based CLI that does some of this: https://github.com/tgree/stm_layout

> And another point is that the the STM32 devices do not have a crossbar connect, thus you cannot connect every pin to any signal. So you have no choice but to look into the datasheet, or use the graphical CubeMX to configure your signals.

### PR DESCRIPTION
**Pull request by henrikssn**
Goals:
- Compilaton errors are understandable (contain necessary information about the context that lead to the error).
- Completely in C++ templates, allows peripherals to perform further validation.
- Allows the user to select the intended peripheral signal (e.g. TX, SCK, MOSI, SDA)
- Allows peripherals to take arguments in arbitrary order (my gift to @salkinium )
- Deduces the pin function at compile time
- Deduces the Peripheral pad at compile time (e.g. Sercom Pad)
- As close to the current signals API as possible

Fixes #481.

### CODE REVIEW

**Comment by salkinium** [comment]
This looks great on first viewing! The STM32 API was still C++11 based, which didn't have variadic templates, so we generated everything with Python. I've always been busy fixing lbuild/modm-devices/tooling stuff that I don't have enough time to properly engage on the C++17 API front. There's a lot more to improve.

I'm still busy with TinyUSB and then I'd like to review/merge your OTA code, so I'll check this out, but probably not fast.

PS: Please fix your formatting style, or I'm going to have a hard time reviewing this.

**Comment by henrikssn** [comment]
> This looks is great on first viewing! The STM32 API was still C++11 based, which didn't have variadic templates, so we generated everything with Python. I've always been busy fixing lbuild/modm-devices/tooling stuff that I don't have enough time to properly engage on the C++17 API front. There's a lot more to improve.
> 
> I'm still busy with TinyUSB and then I'd like to review/merge your OTA code, so I'll check this out, but probably not fast.

Yeah I learned a lot about C++17 while doing this, which was my main motivation. It believe it should be fairly straight-forward to make the STM32 API follow the pattern, but I'd like to make that a separate PR.

The OTA code is still WIP and not ready to merge, I would recommend you start here and then look at the fiber PR.

> PS: Please fix your formatting style, or I'm going to have a hard time reviewing this.

Fixed, I also added the .clang-format file which I used so that I can stay more consistent in the future.

PS: How I dislike those hard tabs...

File: src/modm/board/feather_m0/board.hpp
Comment (salkinium) [inline src/modm/board/feather_m0/board.hpp:104]: This does not have the same semantics. `Leds::write(uint32_t)` accepts a bit vector, `Led::write(bool)` only accepts a single bit.

File: src/modm/board/feather_m0/board.hpp
Comment (salkinium) [inline src/modm/board/feather_m0/board.hpp]: I'd prefer to have Pin and Signal name in the same line.
```suggestion
	Uart0::connect<Rx::Rx, Tx::Tx>();
```

File: src/modm/board/feather_m0/board.hpp
Comment (salkinium) [inline src/modm/board/feather_m0/board.hpp]: Hm?

File: src/modm/platform/gpio/sam/config.hpp.in
Comment (salkinium) [inline src/modm/platform/gpio/sam/config.hpp.in]: Should be generated?

File: src/modm/platform/gpio/sam/pin.hpp
Comment (salkinium) [inline src/modm/platform/gpio/sam/pin.hpp:54]: Shouldn't there be more?

**Review by salkinium — COMMENTED** [review]
I like this very, very much! It's so much cleaner and much faster to generate/compile than the STM32 solution, which generates 200+ files just for the GPIOs, which is just dumb.

I added a closer-to-modm-style clang format, but I guess we're going to have to use `// clang-format off` for some of the short function stuff. You still have to set hardtabs in your editor for the template stuff. Theres a `.editorconfig` for that purpose, it's supposed to be editor independent.

I would still like to see an alias for `using GpioA00 = Gpio<A00>;` for backwards compatiblity? Or perhaps just for STM32? I'm not sure I like having a `A00` type floating around globally (due to `using namespace modm::platform;`), so perhaps it should be `using GpioA00 = Gpio<GpioConfigA00>;`?

**Review by henrikssn — COMMENTED** [review]
> I like this very, very much! It's so much cleaner and much faster to generate/compile than the STM32 solution, which generates 200+ files just for the GPIOs, which is just dumb.

Thanks for the review!
> 
> I added a closer-to-modm-style clang format, but I guess we're going to have to use `// clang-format off` for some of the short function stuff. You still have to set hardtabs in your editor for the template stuff. Theres a `.editorconfig` for that purpose, it's supposed to be editor independent.

I disagree your clang-format is substantially closer to modm-style than the one I added. I force-pushed the branch back to where it originally was, in the future, please make changes in commits so that we can review the differences. I spent multiple hours aligning it with the modm style so it would be good to compare our efforts...

Also feel free to comment on the style in this PR, then we can see how we can fix that in the clang-format file.

I used this command to assess quality of the clang-format file: `find src/modm | grep -e "\.\(c\|h\|hpp\|cpp\)\$" | xargs clang-format -n 2>&1 | wc -l` (lower is better)

Your config: 143759
My config: 143800

Note that my config does allow the short functions, so no need for the `// clang-format off`.

> 
> I would still like to see an alias for `using GpioA00 = Gpio<A00>;` for backwards compatiblity? Or perhaps just for STM32? 

I'm fine with adding the using declarations, but it should be a separate templated header file (which we could probably share between the platforms).

> I'm not sure I like having a `A00` type floating around globally (due to `using namespace modm::platform;`), so perhaps it should be `using GpioA00 = Gpio<GpioConfigA00>;`

How about `PinA01` then? `GpioConfigA00` is too long for my taste.

File: src/modm/board/feather_m0/board.hpp
Comment (henrikssn) [reply src/modm/board/feather_m0/board.hpp:104]: Yes, but a bit vector is implicitly convertible to a bool, so it should be ok in all cases.

Here is the line I am trying to make it work with: https://github.com/modm-io/modm/blob/3072005ddf57cad660a095335573a7614bf533af/src/modm/board/board.cpp.in#L50

Is it expected that it only will blink the first led?

File: src/modm/board/feather_m0/board.hpp
Comment (henrikssn) [reply src/modm/board/feather_m0/board.hpp]: But they are totally different, for example you can use the `Rx` signal from the `Tx` pin (`Tx::Rx`) if we make that decision, which makes no sense to me.

The problem is that the signal to pin mapping is much more flexible on SAM, so we need to be a bit careful keeping them apart.

File: src/modm/platform/gpio/sam/config.hpp.in
Comment (henrikssn) [reply src/modm/platform/gpio/sam/config.hpp.in]: Yes, although I couldn't think of a SAM device with only one port (and afaik there is none with more than 2)

File: src/modm/platform/gpio/sam/pin.hpp
Comment (henrikssn) [reply src/modm/platform/gpio/sam/pin.hpp:54]: We don't have any others implemented for SAM atm. I'll work on extending this list next.

File: src/modm/board/feather_m0/board.hpp
Comment (salkinium) [inline src/modm/board/feather_m0/board.hpp]: This brace should be in the next line.

File: src/modm/platform/gpio/sam/config.hpp.in
Comment (salkinium) [inline src/modm/platform/gpio/sam/config.hpp.in]: This file needs to be converted to tabs and the { put on a newline.

File: examples/samd/interrupt/main.cpp
Comment (salkinium) [inline examples/samd/interrupt/main.cpp]: { on new line.

**Comment by salkinium** [comment]
I'm currently super busy with manufacturing 100 railway signals, so I'm AFK for a few weeks.
If you can fix the { on the newline, then I'm happy to merge this.

File: src/modm/board/feather_m0/board.hpp
Comment (henrikssn) [reply src/modm/board/feather_m0/board.hpp]: Nice catch, done (in .clang-format)

File: examples/samd/interrupt/main.cpp
Comment (henrikssn) [reply examples/samd/interrupt/main.cpp]: Done (also in .clang-format)

File: src/modm/platform/gpio/sam/config.hpp.in
Comment (henrikssn) [reply src/modm/platform/gpio/sam/config.hpp.in]: Braces done, I still need to figure out how to configure my editor (VSCode) for properly emitting tabs

Comment (salkinium) [reply src/modm/platform/gpio/sam/config.hpp.in]: I've added a `.editorconfig` to make GitHub render tabs as 4 spaces. Perhaps you can use the same mechanism. Something like this perhaps:
https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig

**Review by salkinium — APPROVED** [review]
Need to make the CI pass and rebase and squash onto develop please.

File: src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp
Comment (salkinium) [inline src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp:7]: And what did you change? This giant diff really isn't selling me on using clang format and basically just demotivates me to maintain modm entirely. Like, how do you expect me to review this?

File: src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp
Comment (salkinium) [inline src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp]: How in hell is this more readable?

File: src/modm/platform/i2c/bitbang/bitbang_i2c_master_impl.hpp
Comment (salkinium) [inline src/modm/platform/i2c/bitbang/bitbang_i2c_master_impl.hpp]: Srsly? No.

**Comment by salkinium** [comment]
Perhaps using clang format is ok on new files, but we're not reformatting entire swathes of modm, just because one line changes. I'm not wasting my free time fighting an autoformatter.

**Comment by salkinium** [comment]
I'm sorry I'm exploding at you, but these are exactly the same issues that I first encountered ~3 years ago when investigating using clang format for modm (during the porting from xpcc) and it's very frustating to see all the failing corner cases. I like the idea, but I'm going to have to deal with the fallout of this decision for many, many years going forward and I don't want to burn out on this project from preventable causes.

**Comment by salkinium** [comment]
And why did the STM32 CMSIS header and FreeRTOS submodules change?

**Comment by salkinium** [comment]
\*challening the inner grumpy old maintainer\* get off my lawn!1!!

File: src/modm/platform/gpio/sam/config.hpp.in
Comment (henrikssn) [reply src/modm/platform/gpio/sam/config.hpp.in]: I tried it but it makes it impossible to edit the python files, as there we use spaces and python considers the difference important

Comment (salkinium) [reply src/modm/platform/gpio/sam/config.hpp.in]: Then we need to add a filter for python files to the config.

**Comment by henrikssn** [comment]
Rebased + fixed CI

**Comment by henrikssn** [comment]
> I'm sorry I'm exploding at you, but these are exactly the same issues that I first encountered ~3 years ago when investigating using clang format for modm (during the porting from xpcc) and it's very frustating to see all the failing corner cases. I like the idea, but I'm going to have to deal with the fallout of this decision for many, many years going forward and I don't want to burn out on this project from preventable causes.

If I look at the changes, there are more cases where the existing style is inconsistent with the modm style than there are cases where clang-format IMO is wrong.

For example, the contribution guide specifies no space after template keyword, yet that is the case in all of the bitbang implementations. Also, brace wrapping of if/else conditions is done very inconsistently (and this one isn't well defined in the style guide). At some point, you need to bite the apple and decide which one is actually correct.

Obviously, if the goal is to have inconsistent style everywhere, then we should not use clang format _[grumpy new contributor mode on] but then I expect no more comments on the style of my commits... [/grumpy]_

File: src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp
Comment (henrikssn) [reply src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp]: I believe this is caused by indenting the line wrap 4 tabs (should only be one). Let me fix that...

File: src/modm/platform/i2c/bitbang/bitbang_i2c_master_impl.hpp
Comment (henrikssn) [reply src/modm/platform/i2c/bitbang/bitbang_i2c_master_impl.hpp]: Done

**Comment by salkinium** [comment]
> if the goal is to have inconsistent style everywhere

The goals are in order:

1. Don't burn out the maintainer.
2. Have somewhat structured, reviewable PRs, so we don't burn out the maintainer.
3. Have somewhat consistent, readable code, so we don't burn out the maintainer.
4. Don't burn out the maintainer.
 
All are pretty subjective, but forcing me to *search* the diff in a reformatted file is not acceptable. I also contribute to other projects and I don't reformat entirely unrelated lines of code because I want my changes merged and not bikeshed about code formatting.

Our formatting guide is basically about quickly grokking code, by visually matching vertically scoped code elements, like most control structured and function declarations.

1. Use tabs, yes we hate it too, but it's too late to change.
2. Put braces in a new line unless it becomes an unreadable mess.
3. Put the function name on a newline unless it becomes an unreadable mess.
4. Limit yourself to <100 chars unless it becomes an unreadable mess.

This is a cooperation between you and me, not a negotiation. You adapt your style a little and I look the other way and then we can both be maximize our happiness. This has worked quite well for most contributors and it can work well for us too.

> the contribution guide specifies no space after template keyword, yet that is the case in all of the bitbang implementations.

The bit-bang implementation is adapted from code from 2006 I believe. Much modm code is over a decade old. When we ported xpcc to modm, we experimented with clang-format to change the tabs to spaces, but it was just too many changes it would have double the repository size and made it impossible to use git blame for anything anymore.

Which is why using clang-format on purely added files is very acceptable to me, but let's not plough over the entire repo on principle.

File: src/modm/platform/gpio/sam/config.hpp.in
Comment (salkinium) [reply src/modm/platform/gpio/sam/config.hpp.in]: Try adding this to `.editorconfig`, that should make the tabs go away for the Python files.
```
[*.{lb,py}]
indent_style = space
indent_size = 4
```

**Comment by henrikssn** [comment]
I reverted the style changes to the bitbang* files and rebased/squashed against develop.

File: src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp
Comment (salkinium) [reply src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp:7]: Much, much better now. Thank you!

**Review by salkinium — APPROVED** [review]
Excellent.

File: .editorconfig
Comment (rleh) [inline .editorconfig:16]:
```suggestion
indent_style = tab
```

File: src/modm/board/feather_m0/board.hpp
Comment (rleh) [inline src/modm/board/feather_m0/board.hpp]: Looks unintentional to me.

File: src/modm/board/feather_m0/board.hpp
Comment (rleh) [reply src/modm/board/feather_m0/board.hpp:104]: What is the benefit of removing the `SoftwareGpioPort<>` here?
I would like to keep it in order to have the same `board.hpp` files for further automation and testing in the future (#359).

File: src/modm/platform/gpio/sam/module.md
Comment (rleh) [inline src/modm/platform/gpio/sam/module.md]:
```suggestion
- Passing a pin which has a connection to the driver but not the correct instance
```

File: src/modm/platform/gpio/sam/unused.hpp
Comment (rleh) [inline src/modm/platform/gpio/sam/unused.hpp:71]: Why are all the functions removed from `GpioUnused`?

File: src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp
Comment (rleh) [inline src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp:59]: As far as I can see, that's not really necessary, is it?

File: src/modm/platform/uart/sam/uart.hpp.in
Comment (rleh) [inline src/modm/platform/uart/sam/uart.hpp.in:45]: Nice to know! :+1:

File: src/modm/platform/usb/sam/usb.hpp
Comment (rleh) [inline src/modm/platform/usb/sam/usb.hpp]: Why not use the trick from above here too?
```suggestion
		template<class DpPin, class DmPin>
		struct ValidateNotSame {
			static_assert(
				!std::is_same_v<typename DpPin::Base, typename DmPin::Base>,
				"Dp and Dm cannot use the same pin!");
		};
```

**Review by rleh — CHANGES_REQUESTED** [review]
Nice work! :+1: 

I have commented a few remarks and questions inline.

File: .editorconfig
Comment (rleh) [reply .editorconfig:16]: > `[*.{lb,py}]`

Nevermind...

File: src/modm/platform/gpio/sam/unused.hpp
Comment (salkinium) [reply src/modm/platform/gpio/sam/unused.hpp:71]: The SAM port has its own separate peripheral for external interrupts, thus this API is not part of the (platform-specific) GPIO API.

File: src/modm/board/feather_m0/board.hpp
Comment (henrikssn) [reply src/modm/board/feather_m0/board.hpp]: Fixed

File: src/modm/board/feather_m0/board.hpp
Comment (henrikssn) [reply src/modm/board/feather_m0/board.hpp:104]: It wasn't implemented on SAM properly and the new `Gpio<>` is drop-in compatible with a `SoftwareGpioPort<>` that only contains one port (which is the case for all SAM boards at the moment).

Maybe we can move the `SoftwareGpioPort<>` to a common library after stm32/avr supports the same signals API?

File: src/modm/platform/gpio/sam/module.md
Comment (henrikssn) [reply src/modm/platform/gpio/sam/module.md]: Done

File: src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp
Comment (henrikssn) [reply src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp:59]: It is needed since `Gpio` is now a template, so it can't be used unspecialized like this anymore.

It makes sense as different `Gpio` can have different options for `InputType` / `OutputType` (e.g. not all pins supports pull-up/pull-down).

File: src/modm/platform/usb/sam/usb.hpp
Comment (henrikssn) [reply src/modm/platform/usb/sam/usb.hpp]: We require `DpPin` to use the `Dp` peripheral pin and `DmPin` to use `Dm`, so it can't happen that the pins use the same peripheral pin.

This is a problem on SAM UART because there the same signal (e.g. `Pad0`) can be used as both `Rx` and `Tx` peripheral pins.

**Review by henrikssn — COMMENTED** [review]
Thanks for the review!

**Review by rleh — APPROVED** [review]

**Comment by rleh** [comment]
Merge @salkinium?

File: tools/scripts/clang-format.sh
Comment (salkinium) [inline tools/scripts/clang-format.sh:11]: Added a filter so that only newly added files get formatted, otherwise I go crazy as discussed ;-P

**Review by salkinium — APPROVED** [review]
Squashed and rebased, I'll merge once the CI passes.

File: tools/scripts/clang-format.sh
Comment (henrikssn) [reply tools/scripts/clang-format.sh:11]: Thanks! I realized this command is even better, as it reformats only changed lines: `git clang-format upstream/develop`

**Pull request merged**
